### PR TITLE
[Feat] Init Mosaic Augmentation

### PR DESCRIPTION
**Pull request by shijianjian**
Adding mosaic augmentation for a better YOLO-series support. The codebase is expected to be finished by 20th May.

Details of the augmentation may refer to [here](https://programming.vip/docs/yolov5-data-augmentation.html).

Test code:
```python
import torch
import torch.nn as nn
from PIL import ImageDraw
from skimage import io
from kornia.augmentation._2d.mix.mosaic import RandomMosaic
from torchvision.transforms import ToPILImage


def plot(image, box):
    img = ToPILImage()(image)
    img_d = ImageDraw.Draw(img)
    for i, (a, b, c, d) in enumerate(box.numpy()):
        shape = [(a, b), (c, d)]
        img_d.rectangle(shape, outline="red")
    return img


png1 = io.imread("./panda.jpg")
png2 = io.imread("./img.jpeg")

x = torch.cat([
    nn.functional.interpolate(torch.from_numpy(png1).permute(2, 0, 1)[None], (224, 224)),
    nn.functional.interpolate(torch.from_numpy(png2).permute(2, 0, 1)[None], (224, 224)),
]).repeat(4, 1, 1, 1)
boxes = torch.tensor([
    # image 1
    [
        [70., 5, 150, 100],  # head
        [60, 180, 175, 220],  # feet
    ],
    # image 2
    [
        [75, 30, 175, 140],  # head
        [0, 0, 0, 0]  # placeholder
    ]
]).repeat(4, 1, 1)

y, y1 = RandomMosaic(
    (300, 300),
    mosaic_grid=(2, 2), start_ratio_range=(0.3, 0.5), p=1., min_bbox_size=300,
    data_keys=['input', 'bbox_xyxy']
)(x / 255., boxes)
print(y.shape, y1.to_tensor("xyxy")[0])
ToPILImage()(y[0])
plot(y[0], y1.to_tensor("xyxy")[0])
```
output

/cc @hhaAndroid

### CODE REVIEW

File: kornia/augmentation/_2d/mix/cutmix.py [hunk 1]
```diff
@@ -133,3 +133,124 @@ def apply_transform(  # type: ignore
...
+
+    The function returns (inputs, labels), in which the inputs is the tensor that contains the mixup images
+    while the labels is a :math:`(\text{num_mixes}, B, 3)` tensor that contains (label_permuted_batch, lambda)
+    for each cutmix.
+
+    The implementation referred to the following repository: `https://github.com/clovaai/CutMix-PyTorch
+    <https://github.com/clovaai/CutMix-PyTorch>`_.
+
+    Args:
+        height (int): the width of the input image.
```
Comment (edgarriba) [inline kornia/augmentation/_2d/mix/cutmix.py:154]:
```suggestion
        height: the width of the input image.
```

Comment (edgarriba) [reply kornia/augmentation/_2d/mix/cutmix.py:154]: This is automated

File: kornia/augmentation/_2d/mix/cutmix.py [hunk 2]
```diff
@@ -133,3 +133,124 @@ def apply_transform(  # type: ignore
...
+        if height is not None or width is not None:
+            warnings.warn(
+                "height and width can be inferred automatically now. "
+                "The height and width arguments will be removed finally.",
+                category=DeprecationWarning,
+            )
+        self._param_generator = cast(rg.CutmixGenerator, rg.CutmixGenerator(cut_size, beta, num_mix, p=p))
+
+    def transform_labels(
+        self, label: torch.Tensor, params: Dict[str, torch.Tensor]
```
Comment (edgarriba) [inline kornia/augmentation/_2d/mix/cutmix.py:219]:
```suggestion
        self, label: Tensor, params: Dict[str, Tensor]
```
Now you can do `from kornia.core import Tensor`

File: kornia/augmentation/_2d/mix/cutmix.py [hunk 3]
```diff
@@ -133,3 +133,124 @@ def apply_transform(  # type: ignore
...
+        self, input: torch.Tensor, params: Dict[str, torch.Tensor]  # type: ignore
+    ) -> torch.Tensor:
+        height, width = input.size(2), input.size(3)
+        num_mixes = params["mix_pairs"].size(0)
+        batch_size = params["mix_pairs"].size(1)
+
+        _shape_validation(params["mix_pairs"], [num_mixes, batch_size], "mix_pairs")
+        _shape_validation(params["crop_src"], [num_mixes, batch_size, 4, 2], "crop_src")
+
+        out_inputs = input.clone()
```
Comment (edgarriba) [inline kornia/augmentation/_2d/mix/cutmix.py:249]: Won’t this be slow doing a copy ?

File: kornia/constants.py [hunk 4]
```diff
@@ -61,3 +61,4 @@ class DataKey(ConstantBase, Enum, metaclass=EnumMetaFlags):
     BBOX_XYXY = 3
     BBOX_XYWH = 4
     KEYPOINTS = 5
+    TAG = 6
```
Comment (edgarriba) [inline kornia/constants.py:64]: Can you add a comment with a small definition of what TAG means ?

**Review by edgarriba — COMMENTED** [review]
As overall idea, what’s the main difference between v1 and v2 ?

File: kornia/constants.py (hunk 4 above)
Comment (shijianjian) [reply kornia/constants.py:64]: TAG means the class labels for classification.

File: kornia/augmentation/_2d/mix/cutmix.py (hunk 3 above)
Comment (shijianjian) [reply kornia/augmentation/_2d/mix/cutmix.py:249]: Or we may offer an `inplace` argument to allow inplace permutation.

**Comment by shijianjian** [comment]
> As overall idea, what’s the main difference between v1 and v2 ?

`V1` forces the input as `forward(input, label)` which is pretty strict to use. `v2` will follow the new API to have `transform_label`, `transform_mask` to give more flexibility to the end users.

File: kornia/augmentation/_2d/mix/base.py [hunk 5]
```diff
@@ -104,3 +108,162 @@ def forward(  # type: ignore
...
+
+    def apply_func(  # type: ignore
+        self, in_tensor: Tensor, params: Dict[str, Tensor],
+        flags: Optional[Dict[str, Any]] = None
+    ) -> Tensor:
+        if flags is None:
+            flags = self.flags
+        to_apply = params['batch_prob']
+
+        # if no augmentation needed
```
Comment (edgarriba) [inline kornia/augmentation/_2d/mix/base.py:161]:
```suggestion
        # if no augmentation needed
        maybe_apply = to_apply.sum()
        if maybe_apply == 0:
```
to avoid computing sum twice in the elif branch

File: kornia/augmentation/_2d/mix/base.py [hunk 6]
```diff
@@ -104,3 +108,162 @@ def forward(  # type: ignore
...
+    ) -> None:
+        super().__init__(p, p_batch=p_batch, same_on_batch=same_on_batch, keepdim=keepdim)
+        self.data_keys = [DataKey.get(inp) for inp in data_keys]
+
+    def transform_tensor(self, input: Tensor) -> Tensor:
+        """Convert any incoming (H, W), (C, H, W) and (B, C, H, W) into (B, C, H, W)."""
+        _validate_input_dtype(input, accepted_dtypes=[torch.float16, torch.float32, torch.float64])
+        return _transform_input(input)
+
+    def apply_transform(  # type: ignore
```
Comment (edgarriba) [inline kornia/augmentation/_2d/mix/base.py:140]: why all the type ignores? might be worth to specify the `type` of ignore

File: kornia/augmentation/_2d/mix/base.py [hunk 7]
```diff
@@ -104,3 +108,162 @@ def forward(  # type: ignore
...
+    def transform_mask(
+        self, input: Tensor, params: Dict[str, Tensor], flags: Dict[str, Any]
+    ) -> Tensor:
+        return self.apply_transform_mask(input, params, flags)
+
+    def transform_boxes(
+        self, input: Union[Tensor, Boxes], params: Dict[str, Tensor], flags: Dict[str, Any]
+    ) -> Tensor:
+        # input is BxNx4x2 or Boxes.
+        if isinstance(input, Tensor):
```
Comment (edgarriba) [inline kornia/augmentation/_2d/mix/base.py:192]: use `KORNIA_CHECK`

File: kornia/augmentation/_2d/mix/cutmix.py [hunk 8]
```diff
@@ -133,3 +133,124 @@ def apply_transform(  # type: ignore
...
+
+    The implementation referred to the following repository: `https://github.com/clovaai/CutMix-PyTorch
+    <https://github.com/clovaai/CutMix-PyTorch>`_.
+
+    Args:
+        height: the width of the input image.
+        width: the width of the input image.
+        p: probability for applying an augmentation to a batch. This param controls the augmentation
+                   probabilities batch-wisely.
+        num_mix: cut mix times. Default is 1.
```
Comment (edgarriba) [inline kornia/augmentation/_2d/mix/cutmix.py:158]:
```suggestion
        num_mix: cut mix times.
```
defaults are also inlcuded

File: kornia/augmentation/_2d/mix/cutmix.py [hunk 9]
```diff
@@ -133,3 +133,124 @@ def apply_transform(  # type: ignore
...
+            This flag will not maintain permutation order. Default: False.
+        keepdim: whether to keep the output shape the same as input (True) or broadcast it
+                        to the batch form (False). Default: False
+
+    Inputs:
+        - Input image tensors, shape of :math:`(B, C, H, W)`.
+        - Raw labels, shape of :math:`(B)`.
+
+    Returns:
+        Tuple[torch.Tensor, torch.Tensor]:
```
Comment (edgarriba) [inline kornia/augmentation/_2d/mix/cutmix.py:173]:
```suggestion
        Tuple[Tensor, Tensor]:
```

File: kornia/augmentation/_2d/mix/cutmix.py [hunk 10]
```diff
@@ -133,3 +133,124 @@ def apply_transform(  # type: ignore
...
+                  [0.4594, 0.1756, 0.9492]]]]), tensor([[[0.0000, 1.0000, 0.4444],
+                 [1.0000, 0.0000, 0.4444]]]))
+    """
+
+    def __init__(
+        self,
+        height: Optional[int] = None,
+        width: Optional[int] = None,
+        num_mix: int = 1,
+        cut_size: Optional[Union[torch.Tensor, Tuple[float, float]]] = None,
```
Comment (edgarriba) [inline kornia/augmentation/_2d/mix/cutmix.py:203]:
```suggestion
        cut_size: Optional[Union[Tensor, Tuple[float, float]]] = None,
```

File: kornia/augmentation/_2d/mix/cutmix.py [hunk 11]
```diff
@@ -133,3 +133,124 @@ def apply_transform(  # type: ignore
...
+        beta: Optional[Union[torch.Tensor, float]] = None,
```
Comment (edgarriba) [inline kornia/augmentation/_2d/mix/cutmix.py:204]:
```suggestion
        beta: Optional[Union[Tensor, float]] = None,
```

File: kornia/augmentation/_2d/mix/cutmix.py [hunk 12]
```diff
@@ -133,3 +133,124 @@ def apply_transform(  # type: ignore
...
+        same_on_batch: bool = False,
+        p: float = 1.0,
+        keepdim: bool = False,
+    ) -> None:
+        super().__init__(p=1.0, p_batch=p, same_on_batch=same_on_batch, keepdim=keepdim)
+        if height is not None or width is not None:
+            warnings.warn(
+                "height and width can be inferred automatically now. "
+                "The height and width arguments will be removed finally.",
```
Comment (edgarriba) [inline kornia/augmentation/_2d/mix/cutmix.py:213]:
```suggestion
                "The height and width arguments will be removed in a future version.",
```

File: kornia/augmentation/_2d/mix/mosaic.py [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+            )
+        if flags["cropping_mode"] == "slice":  # uses advanced slicing to crop
+            return crop_by_indices(input, params["src"], flags["output_size"])
+        raise NotImplementedError(f"Not supported type: {flags['cropping_mode']}.")
+
+    def apply_non_transform(
+        self, input: Tensor, params: Dict[str, Tensor], flags: Dict[str, Any]
+    ) -> Tensor:
+        if flags["output_size"] is not None:
+            return interpolate(
```
Comment (edgarriba) [inline kornia/augmentation/_2d/mix/mosaic.py:165]: use `kornia.geometry.transform.resize`

File: kornia/constants.py (hunk 4 above)
Comment (edgarriba) [reply kornia/constants.py:64]: not a very intuitive names compared to the others

File: kornia/augmentation/_2d/mix/mosaic.py [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+
+    Examples:
+        >>> mosaic = RandomMosaic()
+        >>> input = torch.randn(8, 3, 16, 16)
+        >>> mosaic(input).shape
+    """
+
+    def __init__(
+        self,
+        output_size: Optional[Tuple[int, int]] = None,
```
Comment (edgarriba) [inline kornia/augmentation/_2d/mix/mosaic.py:35]: i like parameters that contains None in the last terms. Possibly the last before `data_keys` or `p`

File: kornia/augmentation/_2d/mix/cutmix.py (hunk 3 above)
Comment (edgarriba) [reply kornia/augmentation/_2d/mix/cutmix.py:249]: not sure -- i'd avoid if possible out of place ops

File: kornia/augmentation/_2d/mix/cutmix.py [hunk 15]
```diff
@@ -49,9 +51,9 @@ def cutmix_label(labels, out_labels, size):
         p (float): probability for applying an augmentation to a batch. This param controls the augmentation
                    probabilities batch-wisely.
         num_mix (int): cut mix times. Default is 1.
-        beta (float or torch.Tensor, optional): hyperparameter for generating cut size from beta distribution.
+        beta (float or Tensor, optional): hyperparameter for generating cut size from beta distribution.
```
Comment (edgarriba) [inline kornia/augmentation/_2d/mix/cutmix.py:54]:
```suggestion
        beta: hyperparameter for generating cut size from beta distribution.
```
not needed

File: test/augmentation/test_container.py [hunk 16]
```diff
@@ -297,6 +297,22 @@ def test_random_flips(self, device, dtype):
...
+        crop_width, crop_height = 3, 3
+        input = torch.randn(3, 3, width, height, device=device, dtype=dtype)
+        bbox = torch.tensor(
+            [[[1.0, 1.0, 2.0, 2.0], [0.0, 0.0, 1.0, 2.0], [0.0, 0.0, 2.0, 1.0]]], device=device, dtype=dtype
+        ).expand(3, -1, -1)
+        aug = K.AugmentationSequential(
+            K.RandomCrop((crop_width, crop_height), padding=1, cropping_mode='resample', fill=0),
+            K.RandomHorizontalFlip(p=1.0),
+            K.RandomMosaic(p=1.),
+            data_keys=["input", "bbox_xyxy"],
```
Comment (edgarriba) [inline test/augmentation/test_container.py:311]: @shijianjian what happens with the transformation matrix after ? is it composed too ?

Comment (shijianjian) [reply test/augmentation/test_container.py:311]: Mix-series shall not support transformation matrix I think. An error/warning shall be raised when access transformation matrix and performing inverse.

**Comment by edgarriba** [comment]
@shijianjian any update here ?

**Comment by shijianjian** [comment]
> @shijianjian any update here ?

Did not notice there are bugs there... I will finish this weekend.

**Comment by edgarriba** [comment]
@shijianjian could you rebase and fix the mypy issues ? want to merge this ASAP this week for the release

File: test/geometry/test_boxes.py [hunk 17]
```diff
@@ -165,6 +165,8 @@ def test_boxes_to_tensor(self, shape: Tuple[int], device, dtype):
         assert boxes_vertices_plus.shape == expected_vertices_plus.shape  # type: ignore
         assert_allclose(boxes_vertices_plus, expected_vertices_plus)
 
+    # TODO (jian): verify later
+    @pytest.mark.skip(reason="For some reason this test crash with 'vertices', 'vertices_plus'")
```
Comment (edgarriba) [inline test/geometry/test_boxes.py:169]: @shijianjian the `Boxes` validation test must be updated

**Comment by edgarriba** [comment]
@shijianjian i updated the snippet using kornia functions and seems alright except for the `output_size` param

```python
import torch
import cv2

import kornia as K
from kornia.augmentation import RandomMosaic


def plot(img, box):
    img_vis = img.clone()
    img_vis = K.utils.draw_rectangle(
        img_vis, box, color=torch.tensor([255,0,0]))
    cv2.imshow("mosaic", K.tensor_to_image(img_vis))
    cv2.waitKey(0)


img1 = K.io.load_image("/home/edgar/Downloads/1656581296989.jpg", K.io.ImageLoadType.RGB32)
img2 = K.io.load_image("/home/edgar/Downloads/IMG_20220617_193701530.jpg", K.io.ImageLoadType.RGB32)

x = K.core.concatenate([
    K.geometry.resize(img1[None], (224, 224)),
    K.geometry.resize(img2[None], (224, 224))
]).repeat(4, 1, 1, 1)

boxes = torch.tensor([
    # image 1
    [
        [70., 5, 150, 100],  # head
        [60, 180, 175, 220],  # feet
    ],
    # image 2
    [
        [75, 30, 175, 140],  # head
        [0, 0, 0, 0]  # placeholder
    ]
]).repeat(4, 1, 1)

plot(x[:1], boxes[:1])

aug = RandomMosaic(
    (224, 224),
    mosaic_grid=(2, 2), start_ratio_range=(0.3, 0.5), p=1., min_bbox_size=300,
    data_keys=['input', 'bbox_xyxy'])

y, y1 = aug(x, boxes)

plot(y[:1], y1[:1])
```

**Comment by shijianjian** [comment]
@edgarriba Finally, the annoying bug has been fixed. XD. Safe to merge I suppose.

**Comment by edgarriba** [comment]
cool -- can we add somewhere a usage example or colab tutorials. This is a great feature to promote

**Comment by shijianjian** [comment]
Sure thing. I would prepare one.

**Comment by Twsaiae** [comment]
I have a question, which is how do the category information of labels correspond when using Mosaic enhancement on object detection datasets? What is the category corresponding to each label in the generated graph?
